DB/src/main/java/io/deephaven/db/v2/sources/regioned/RegionedColumnSourceWithDictionary.java

```diff
@@ -0,0 +1,299 @@
+        super(dataType, componentType);
+    }
+
+    @Override
+    public <ALTERNATE_DATA_TYPE> boolean allowsReinterpret(@NotNull Class<ALTERNATE_DATA_TYPE> alternateDataType) {
+        return alternateDataType == long.class || super.allowsReinterpret(alternateDataType);
+    }
+
+    @Override
+    protected <ALTERNATE_DATA_TYPE> ColumnSource<ALTERNATE_DATA_TYPE> doReinterpret(@NotNull Class<ALTERNATE_DATA_TYPE> alternateDataType) {
+        //noinspection unchecked
+        return alternateDataType == long.class ? (ColumnSource<ALTERNATE_DATA_TYPE>) new AsLong() : super.doReinterpret(alternateDataType);
+    }
+
+    @Override
+    public void releaseCachedResources() {
+        super.releaseCachedResources();
+    }
+
+    private final class AsLong
+            extends RegionedColumnSourceBase<Long, DictionaryKeys, ColumnRegionLong<DictionaryKeys>>
+            implements ColumnSourceGetDefaults.ForLong {
+
+        private final ColumnRegionLong<DictionaryKeys> nullRegion;
+        private volatile ColumnRegionLong<DictionaryKeys>[] wrapperRegions;
+
+        private AsLong() {
+            super(long.class);
+            nullRegion = ColumnRegionLong.createNull(PARAMETERS.regionMask);
+            //noinspection unchecked
+            wrapperRegions = new ColumnRegionLong[0];
+        }
+
+        @Override
+        public long getLong(final long elementIndex) {
+            return (elementIndex == NULL_KEY ? getNullRegion() : lookupRegion(elementIndex)).getLong(elementIndex);
+        }
+
+        @Override
+        public int addRegion(@NotNull final ColumnDefinition<?> columnDefinition, @NotNull final ColumnLocation columnLocation) {
+            return RegionedColumnSourceWithDictionary.this.addRegion(columnDefinition, columnLocation);
+        }
+
+        @Override
+        <OTHER_REGION_TYPE> int addRegionForUnitTests(@NotNull final OTHER_REGION_TYPE region) {
+            return RegionedColumnSourceWithDictionary.this.addRegionForUnitTests(region);
+        }
+
+        @NotNull
+        @Override
+        ColumnRegionLong<DictionaryKeys> getNullRegion() {
+            return nullRegion;
+        }
+
+        @Override
+        public int getRegionCount() {
+            return RegionedColumnSourceWithDictionary.this.getRegionCount();
+        }
+
+        @Override
+        public ColumnRegionLong<DictionaryKeys> getRegion(final int regionIndex) {
+            final ColumnRegionObject<DATA_TYPE, Values> sourceRegion = RegionedColumnSourceWithDictionary.this.getRegion(regionIndex);
+            if (sourceRegion instanceof ColumnRegion.Null) {
+                return nullRegion;
+            }
+            ColumnRegionLong<DictionaryKeys>[] localWrappers;
+            ColumnRegionLong<DictionaryKeys> wrapper;
+            if ((localWrappers = wrapperRegions).length > regionIndex && (wrapper = localWrappers[regionIndex]) != null) {
```

Comment:
Why do the assignments outside of the synchronized block?

Reply:
Not sure I follow the question. I don't want to read the `volatile wrapperRegions` reference twice, nor do I want to do the array lookup twice. It would *probably be fine* in this case, but it's something I avoid out of habit and hygiene.
I'm reasonably convinced I have the right thread-safety decisions here; by making `localWrappers` volatile I avoid any risk that someone might see an incompletely copied version of the array, and wrapper sources are immutable.

Reply:
This is the expression in the if and also in the later if inside the synchronized:
`(localWrappers = wrapperRegions).length > regionIndex && (wrapper = localWrappers[regionIndex]) != null)`

Outside the synchronized block, if that expression evaluates to true, we bail.  So the assignments to localWrappers and wrapper are useless at that point.  So it is confusing for me why we are doing them.

Reply:
Discussed on the phone.